### PR TITLE
MXTaggedEvents: Expose "m.tagged_events" according to (MSC2437)

### PR DESCRIPTION
**Pull request by yostyle**
MSC: matrix-org/matrix-doc#2437

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -31,6 +31,7 @@ Changes in 0.17.4 (2020-12-02)
 🙌 Improvements
  * Pods: Update JitsiMeetSDK to 2.11.0 to be able to build using Xcode 12.2 (vector-im/element-ios/issues/3808).
  * Pods: Update Realm to 10.1.4 to be able to `pod lib lint` using Xcode 12.2 (vector-im/element-ios/issues/3808).
+ * MXTaggedEvents: Expose "m.tagged_events" according to [MSC2437](https://github.com/matrix-org/matrix-doc/pull/2437).
```
[the remark this answers is not in the document]
Comment (yostyle) [reply CHANGES.rst]: Sorry I updated it

File: MatrixSDK/JSONModels/MXTaggedEvents.m [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        [tagDict removeObjectForKey:eventId];
+    }
+    
+    
+    NSMutableDictionary<NSString*, NSDictionary<NSString*, NSDictionary*>* > *updatedTags = [_tags mutableCopy];
+    if (!updatedTags)
+    {
+        updatedTags = [NSMutableDictionary dictionary];
+    }
+    updatedTags[tag] = tagDict;
```
[the remark this answers is not in the document]
Comment (yostyle) [reply MatrixSDK/JSONModels/MXTaggedEvents.m]: I rewrote this method to improve the flow.

**Review by manuroe — APPROVED** [review]

File: MatrixSDK/Utils/MXTools.m [hunk 3]
```diff
@@ -116,6 +116,7 @@ + (void)initialize
                                 kMXEventTypeStringSecretRequest,
                                 kMXEventTypeStringSecretSend,
                                 kMXEventTypeStringSecretStorageDefaultKey,
+                                kMXEventTypeStringTaggedEvents,
```
Comment (giomfo) [inline MatrixSDK/Utils/MXTools.m:119]: We should remove the last comma

File: MatrixSDK/Data/MXRoomAccountData.m [hunk 4]
```diff
@@ -53,6 +73,7 @@ - (void)encodeWithCoder:(NSCoder *)aCoder
 {
     [aCoder encodeObject:_tags forKey:@"tags"];
     [aCoder encodeObject:_readMarkerEventId forKey:@"readMarkerEventId"];
+    [aCoder encodeObject:_taggedEvents forKey:@"taggedEvents"];
```
[the remark this answers is not in the document]
Comment (yostyle) [reply MatrixSDK/Data/MXRoomAccountData.m:76]: the value may be nil : https://developer.apple.com/documentation/foundation/nskeyedarchiver/1412067-encodeobject

[the remark this answers is not in the document]
Comment (giomfo) [reply MatrixSDK/Data/MXRoomAccountData.m:76]: ok, we may ignore my remark - we already don't check the existing `tags` which is nullable.

File: MatrixSDK/Data/MXRoomAccountData.m [hunk 5]
```diff
@@ -31,12 +31,31 @@ - (void)handleEvent:(MXEvent *)event
...
+             withTag:(NSString*)tag
+{
+    MXTaggedEventInfo *taggedEventInfo;
+    MXJSONModelSetMXJSONModel(taggedEventInfo, MXTaggedEventInfo, _taggedEvents.tags[tag][eventId]);
+    return taggedEventInfo;
+}
+
+- (NSArray<NSString *> *)getTaggedEventsIds:(NSString*)tag
+{
+    return _taggedEvents.tags[tag].allKeys;
```
[the remark this answers is not in the document]
Comment (giomfo) [reply MatrixSDK/Data/MXRoomAccountData.m:56]: we ignore

File: MatrixSDK/JSONModels/MXTaggedEvents.m [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    if (self)
+    {
+        _tags = [aDecoder decodeObjectForKey:@"tags"];
+    }
+    return self;
+}
+
+- (void)encodeWithCoder:(NSCoder *)aCoder
+{
+    [aCoder encodeObject:_tags forKey:@"tags"];
```
[the remark this answers is not in the document]
Comment (giomfo) [reply MatrixSDK/JSONModels/MXTaggedEvents.m:104]: ignored

File: MatrixSDK/JSONModels/MXTaggedEventInfo.m [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+@implementation MXTaggedEventInfo
+
++(id)modelFromJSON:(NSDictionary *)JSONDictionary
+{
+    MXTaggedEventInfo *taggedEventInfo = [[MXTaggedEventInfo alloc] init];
+    if (taggedEventInfo)
+    {
+        MXJSONModelSetArray(taggedEventInfo.keywords, JSONDictionary[kTaggedEventInfoKewordsJSONKey]);
+        MXJSONModelSetUInt64(taggedEventInfo.originServerTs, JSONDictionary[kTaggedEventInfoOriginServerTsJSONKey]);
+        MXJSONModelSetUInt64(taggedEventInfo.taggedAt, JSONDictionary[kTaggedEventInfoTaggedAtJSONKey]);
```
[the remark this answers is not in the document]
Comment (giomfo) [reply MatrixSDK/JSONModels/MXTaggedEventInfo.m:46]: We decided to use kMXUndefinedTimestamp

**Review by giomfo — APPROVED** [review]
LGTM

**Pull request merged**